state/upgrades.go

```diff
@@ -1388,3 +1388,24 @@ func addRelationStatus(st *State) error {
 	})
 	return errors.Annotate(err, "adding relation status")
 }
+
+// CopyMongoSpaceToHASpaceConfig copies the Mongo space name from
+// ControllerInfo to the HA space name in ControllerConfig.
+func CopyMongoSpaceToHASpaceConfig(st *State) error {
+	info, err := st.ControllerInfo()
+	if err != nil {
+		return errors.Annotate(err, "cannot get controller info")
+	}
+	if info.MongoSpaceState != MongoSpaceValid {
+		return nil
+	}
+
+	settings, err := readSettings(st.db(), controllersC, controllerSettingsGlobalKey)
+	if err != nil {
+		return errors.Annotate(err, "cannot get controller config")
+	}
+
+	settings.Set(controller.JujuHASpace, info.MongoSpaceName)
```

Comment:
considering a desire to make things idempotent, should we check to see if something is already set?

what happens if both exist?

I think ultimately that controllerSettingsGlobalKey should win?


We might also want to check for "MongoSpaceValid" but the actual Space value is "".

Reply:
I did think of this; but assuming someone in the field is upgrading a controller to 2.4, the only way there could be a configured HA space would be if they bootstrapped with 2.4, as this is not a stable release. There should be no such config value... or have I got that wrong?